### PR TITLE
ws13: DeepSeek V4 compressed attention implemented and gated (compressor + indexer + integration), plus the arch-gap evidence

### PR DESCRIPTION
**Pull request by bong-water-water-bong**
### **User description**
Scoping only — **no engine code in this PR.** It exists so the next person starts from measurements instead of re-deriving them, and so nobody closes the census alert with an alias.

## Why the flagged families are not aliases
Recorded in `Testing/arch-gaps.md` (lands with #2244), from the official HF configs: V4.1 is 40L/H5120/**384 experts** against V4's 43L/H4096/256 and adds `engram_*`/`candidate_*`/dspark; Mamba-3 carries parameters (`mamba_mimo_rank`, `mamba_bc_norm_eps`, `mamba_rope_fraction`, `mamba_a_floor`) that have no Mamba-2 equivalent; `molmoact2` is a VLA with a flow-matching action expert; `cosmos3edge` is a video world model.

## What is measured here (from `model.safetensors.index.json`, the real tensor inventory)
| | V4-Flash | V4.1-Flash |
|---|---|---|
| tensors / size | 69,187 / **159.6 GB** | 96,085 / **510.3 GB** |
| compress_ratios nonzero | **41/44** {0,4,128} | 38/43 {0,1,2}, but only 4 layers carry compressor tensors, 8 carry indexer tensors |
| fp8 | e4m3, block [128,128] | e4m3 + `expert_dtype: fp4`, block [32,32] |
| V4.1-only | — | `engram.*` (2 layers), `attn.indexer.k_norm/wk`, `ffn.gate.bias_vl` (40), rewritten MTP, vision tower + aligner |
| dropped | — | `compressor.ape`, `ffn.gate.tidNeid` (hash routing), V4-style MTP projections |

The engine's V4 path (`include/deepseek_v4.h`) implements mHC, shared-KV MQA, sinks and MoE; it does **not** implement the CSA/HCA compressors or the Lightning Indexer — the existing gate used sliding-only layers. So a **real V4 checkpoint** needs them too, and one implementation serves both families (V4.1 simply uses fewer such layers).

## The binding constraint
The loader materializes every tensor as f32, so 510 GB fp8 is ~2 TB resident. The workstream therefore takes its weight path from **WS-07** (expert staging/PagedWeight) and **WS-11** (NVMe→DRAM→SRAM) and its format from **WS-05** (1BP v2), instead of inventing a fourth mechanism.

## Phases and gates
P0 instrument first (tiny-config HF oracle; the pattern is `Testing/e2e_gen_check_hf.py`, unusable as-is because it loads a whole checkpoint in fp32) → P1 compressors + indexer + shape-agnostic loader + GGUF aliases, then a **real V4-Flash identity gate** against the independent GGUF reference → P2 V4.1 deltas, a vision scope decision, and the Mamba-3 lane. Honesty-tag policy per WS-00: nothing is claimed without the number.

Also updates `research/PLAN.md`, `research/TRACKING.md` (WS-13 row) and the `research/README.md` tree.

___

### **PR Type**
Enhancement, Tests, Documentation

___

### **Description**
- Implement CSA/HCA compressors and Lightning Indexer in engine

- Parse `layer_types`/`compress_ratios`, add compress-rope theta

- Wire compressed entries into sliding-window attention, plus per-layer state dump

- Add HF-oracle gates, fixture generator and ws13 research docs

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["HF fixture generator<br/>make_mini_deepseek_v41.py"] -- "forward hooks" --> B["compressor_ref / indexer_scores / hidden_states"]
  C["config.json<br/>layer_types, compress_ratios"] -- "parse" --> D["deepseek_v4.cpp<br/>compressor + indexer + attention"]
  B -- "1e-6 oracle" --> E["cmp_deepseek_v4_compressor / _incremental"]
  B -- "score table gate" --> F["cmp_deepseek_v4_indexer_scores.py"]
  D -- "per-layer streams" --> G["cmp_deepseek_v4_layers.py"]
  H["run_all.sh"] -- "builds + runs" --> E
  H -- "builds + runs" --> F
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>deepseek_v4.cpp</strong><dd><code>Implement CSA/HCA compressors, Lightning Indexer and attention </code><br><code>integration</code></dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-97e9f2e1d1790ecd5713e6792dcc6cc8782031a8330de3aa9d1e127a272597e9">+506/-12</a></td>

</tr>

<tr>
  <td><strong>deepseek_v4.h</strong><dd><code>Declare compressor config, params, state and new signatures</code></dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-ab06226750544fd8153609e891331ef5050c8404838941ed57b362ec4660d448">+125/-2</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>9 files</summary><table>
<tr>
  <td><strong>cmp_deepseek_v4_compressor.cpp</strong><dd><code>New gate comparing batched compressor output to HF reference</code></dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-86623e36d5a149b5f4087d5972e8bff56b06a2fcb8888f371848baf5eb9d65d8">+110/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>cmp_deepseek_v4_compressor_incremental.cpp</strong><dd><code>New gate for token-by-token compressor state machine</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-da121dbd238a6a4d1e3c92cd7062d38ca4789a896e959787ca51df69200a8d6c">+97/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>cmp_deepseek_v4_indexer.cpp</strong><dd><code>Indexer comparison harness with score dump and tie analysis</code></dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-ca44136b342a9882187c48279ba9eaada8f345309dad2acb4e8d8803ec68f69a">+161/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>cmp_deepseek_v4_indexer_scores.py</strong><dd><code>Python gate for indexer score table and selection validity</code></dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-6a67f9ecc9e3e4eaffde5e877c7c7bedce27779473fba1af4ed4f1378a890878">+114/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>cmp_deepseek_v4_layers.py</strong><dd><code>Per-layer engine-vs-HF residual stream bound check</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-d3084bcc74324de65c8f1905c65afdfd6464f01c4705837394a86b342e1236db">+82/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>cmp_deepseek_v4.cpp</strong><dd><code>Add optional per-layer state dump output</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-299135e8982a05827eb519ecff427f101d0688f789c1726a160f01e0290feec6">+25/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>make_mini_deepseek_v41.py</strong><dd><code>Generate tiny V4 fixtures with hook-captured oracles</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-082a50188505972f7c280e07563c521fc422c79eee4166d4cdd97690a60e9111">+242/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>make_ws13_fixtures.sh</strong><dd><code>Script generating the seven ws13 fixture profiles</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-caaa3406add622282509c16175df73a1cb3cdbb37631978e12c9866151fae9aa">+40/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>run_all.sh</strong><dd><code>Wire ws13 compressor, indexer and per-layer gates into CI</code></dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-60779953d5d72f0d8ef3aea52f86b0b365bbdb42457aae2e5955d6320e32eb2c">+54/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>7 files</summary><table>
<tr>
  <td><strong>deepseek-v4-flash-reverse-engineering.md</strong><dd><code>Mark report superseded and flag fictional architecture claims</code></dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-4cf8c49ecb0e4291ad16bac3c059b42ef735ee4f84a27ac84bad4322491c178f">+25/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>TRACKING.md</strong><dd><code>Add WS-13 row with measured gate results</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-484bdea931535c909d752bdac46da38ca729de4b22399cc5fe8582b6da2010ed">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>PLAN.md</strong><dd><code>Add WS-13 workstream to the roadmap graph</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-f3f7ddf6a6fbad119bdc11a5f3de7606ae3cbb2809a092d12e286407d0086fe7">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>README.md</strong><dd><code>Index the ws13 arch-gap-closure research folder</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-7457dc94d4573a7b15c63d74287dc45d204bc028ca7a5c11b5b374389e6c34e0">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>FINDINGS.md</strong><dd><code>Record measured gates, closed GGUF route and findings</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-a3b25a364f3b753c94cc328976ba49ede08b4bf3a439fb2682eddf2e716de4b8">+474/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>SPEC-v41-modules.md</strong><dd><code>Specify V4.1-only modules and Mamba-3 lane</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-88e9bab69a9e6cebe28c8689523c6a36f01b64778caf63725a40404444790e97">+115/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>README.md</strong><dd><code>Describe the ws13 workstream scope and layout</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/1bit-MONSTER/1bit-MONSTER/pull/2245/files#diff-c2f5f54483e0c02b3faaa4618f2e77b69000edc27d2c6d5fab3def686dde35ba">+107/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by bong-water-water-bong** [comment]
Second commit: **Phase 0 started**, and it changed what the plan can assume.

**The oracle for the missing machinery already exists.** transformers 5.16.1 (both boxes) ships `DeepseekV4` *with* `compressed_sparse_attention` / `heavily_compressed_attention` layer types, `compress_rates` and `DeepseekV4CSACache` (the indexer) — i.e. the reference implements exactly what `src/deepseek_v4.cpp` lacks. It has **no** `DeepseekV41` class (`deepseek_v41` is absent from `CONFIG_MAPPING`), so V4.1-only modules (engram, `candidate_*`, `ffn.gate.bias_vl`, MTP) have no transformers reference — their authority is DeepSeek's own `inference/*.py`. The old `~/models/venv-zaya` interpreter is gone from both boxes; the generator now runs on the surviving `ft-zaya` venv.

**`Testing/make_mini_deepseek_v41.py`** (new) builds tiny seeded fixtures with two profiles and writes HF logits **plus per-layer hidden states**. Measured with the existing, unmodified `cmp_deepseek_v4.cpp`:

| fixture | window / prompt | compressor entries | engine vs reference | |
|---|---|---|---|---|
| sliding | 32 / 5 | 0 | top1 **342 = 342**, 20/20 | **PASS** (exit 0) |
| compressed | **4** / 64 | CSA 16 | top1 534 ≠ 685, 16/20 | FAIL (exit 1) |
| compressed | **4** / 160 | CSA 40 + HCA 1 | top1 301 ≠ 207, **14/20** | FAIL (exit 1) |

The sliding row reproduces the documented 2026-08-16 mini-gate (`top1 342`, 20/20) exactly, and the divergence grows with how much the compressor matters — the gate is green on what exists and red on what does not.

**Traps found (all recorded in `FINDINGS.md`)**: the first compressed fixture used `sliding_window=32` with a 5-token prompt and the **engine passed it** — a blind gate; `compressor.position_bias` is shaped `[compress_rate, dim]` so the rate is *baked into the tensors* (no same-weights ablation, and a loader must read `compress_ratios`); the transformers names (`compressor.gate_proj`/`kv_proj`/`position_bias`, `compressor.indexer.*`) differ from the published checkpoint's (`attn.compressor.wgate`/`wkv`/`ape`, `attn.indexer.*`) for the same maths; HCA emits nothing below 128 tokens so it needs its own fixture.

**P0.1 is partially done**: the sliding profile passes end to end, but the `≤1e-6` **per-layer** bound is not yet measurable — the engine side has no per-layer dump, so the comparator only sees final logits. That is the next instrument step, then P0.2 (read `inference/engram.py` for the V4.1-only maths).
